beaconChain/src/stateTransition/epoch/variables.ts

```diff
@@ -0,0 +1,111 @@
+      return attestation;
+    }
+  });
+
+  const currentEpochBoundaryAttesterIndices: ValidatorIndex[] = currentEpochBoundaryAttestations
+    .map((attestation: PendingAttestation) => getAttestationParticipants(state, attestation.data, attestation.aggregationBitfield))
+    .reduce((previousValue: ValidatorIndex[], currentValue: ValidatorIndex[]) => previousValue.concat(...currentValue));
+
+  const currentEpochBoundaryAttestingBalance = getTotalBalance(state, currentEpochBoundaryAttesterIndices);
+
+  // Validators attesting during the previous epoch
+  const previousTotalBalance = getTotalBalance(state, getActiveValidatorIndices(state.validatorRegistry, previousEpoch));
+
+  // Validators that made an attestation during the previous epoch, targeting the previous justified slot
+  const previousEpochAttestations: PendingAttestation[] = state.latestAttestations.filter((attestation) => {
+    if (previousEpoch === slotToEpoch(attestation.data.slot)) {
+      return attestation;
+    }
+  });
+
+  const previousEpochAttesterIndices: ValidatorIndex[] = previousEpochAttestations
+    .map((attestation: PendingAttestation) => getAttestationParticipants(state, attestation.data, attestation.aggregationBitfield))
+    .reduce((previousValue: ValidatorIndex[], currentValue: ValidatorIndex[]) => previousValue.concat(...currentValue));
+
+  const previousEpochAttestingBalance: Gwei = getTotalBalance(state, previousEpochAttesterIndices);
+
+  // Validators justifying the epoch boundary block at the start of the previous epoch
+  const previousEpochBoundaryAttestations: PendingAttestation[] = previousEpochAttestations.filter((attestation) => {
+    if (attestation.data.epochBoundaryRoot === getBlockRoot(state, getEpochStartSlot(previousEpoch))) {
+      return attestation;
+    }
+  });
+
+  const previousEpochBoundaryAttesterIndices: ValidatorIndex[] = previousEpochBoundaryAttestations
+    .map((attestation) => getAttestationParticipants(state, attestation.data, attestation.aggregationBitfield))
+    .reduce((previousValue: ValidatorIndex[], currentValue: ValidatorIndex[]) => previousValue.concat(...currentValue));
+
+  const previousEpochBoundaryAttestingBalance: Gwei = getTotalBalance(state, previousEpochBoundaryAttesterIndices);
+
+  // Validators attesting to the expected beacon chain head during the previous epoch
+  const previousEpochHeadAttestations: PendingAttestation[] = previousEpochAttestations.filter((attestation) => {
+    if (attestation.data.beaconBlockRoot === getBlockRoot(state, attestation.data.slot)) {
+      return attestation;
+    }
+  });
+
+  const previousEpochHeadAttesterIndices: ValidatorIndex[] = previousEpochAttestations
+    .map((attestation) => getAttestationParticipants(state, attestation.data, attestation.aggregationBitfield))
+    .reduce((previousValue: ValidatorIndex[], currentValue: ValidatorIndex[]) => previousValue.concat(...currentValue));
```

Comment:
deduplicate
see the comment for `currentEpochBoundaryAttesterIndices` above

Reply:
oh ya it done